public/app/features/variables/testing/adHocVariableBuilder.ts

```diff
@@ -0,0 +1,51 @@
+import { AdHocVariableModel, VariableHide, AdHocVariableFilter } from 'app/features/templating/variable';
+import { v4 } from 'uuid';
+
+const initial: AdHocVariableModel = {
+  uuid: v4(),
+  global: false,
+  type: 'adhoc',
+  name: '',
+  hide: VariableHide.dontHide,
+  label: '',
+  skipUrlSync: false,
+  index: -1,
+  initLock: null,
+  datasource: null,
+  filters: [],
+};
+
+export class AdHocVariableBuilder {
+  private variable: AdHocVariableModel;
+
+  constructor(partial: Partial<AdHocVariableModel>) {
+    this.variable = {
+      ...initial,
+      ...partial,
+    };
+  }
+
+  withName(name: string) {
+    this.variable.name = name;
+    return this;
+  }
+
+  withUUID(uuid: string) {
+    this.variable.uuid = uuid;
+    return this;
+  }
+
+  withDatasource(datasource: string) {
+    this.variable.datasource = datasource;
+    return this;
+  }
+
+  withFilters(filters: AdHocVariableFilter[]) {
+    this.variable.filters = filters;
+    return this;
+  }
+
+  build(): AdHocVariableModel {
+    return this.variable;
+  }
+}
```

Comment:
Hmm, why do we need another builder?

Reply:
We don't really need another but with this pattern we extract shared functionality into a base builder. So we can end up with builders with variable specific functionality which makes them a bit easier.

So basically the same but a bit more flexible when it comes to extending it. If that makes sense?

Reply:
Not a fan of classes though, but I see your point.